src/script/components/MessagesList/VirtualizedMessagesList/useLoadMessages.ts

```diff
@@ -0,0 +1,127 @@
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see http://www.gnu.org/licenses/.
+ *
+ */
+
+import {useCallback, useEffect, useLayoutEffect, useRef} from 'react';
+
+import {Virtualizer} from '@tanstack/react-virtual';
+
+import {ConversationRepository} from 'src/script/conversation/ConversationRepository';
+import {Conversation} from 'src/script/entity/Conversation';
+import {useKoSubscribableChildren} from 'Util/ComponentUtil';
+import {isLastReceivedMessage} from 'Util/conversationMessages';
+
+const SCROLL_THRESHOLD = 10;
+
+interface Props {
+  conversation: Conversation;
+  conversationRepository: ConversationRepository;
+  loadingMessages: boolean;
+  onLoadingMessages: (isLoading: boolean) => void;
+  itemsLength: number;
+}
+
+export const useLoadMessages = (
+  virtualizer: Virtualizer<HTMLDivElement, Element>,
+  {conversation, conversationRepository, loadingMessages, onLoadingMessages, itemsLength}: Props,
+) => {
+  const {isLoadingMessages, hasAdditionalMessages} = useKoSubscribableChildren(conversation, [
+    'isLoadingMessages',
+    'hasAdditionalMessages',
+  ]);
+
+  const nbOfMessages = useRef(0);
+
+  const loadPrecedingMessages = useCallback(async () => {
+    const shouldPullMessages = !isLoadingMessages && hasAdditionalMessages;
+
+    if (shouldPullMessages) {
+      onLoadingMessages(true);
+      await conversationRepository.getPrecedingMessages(conversation).then(() => {
+        onLoadingMessages(false);
+      });
+
+      nbOfMessages.current = itemsLength;
+    }
+  }, [conversation, conversationRepository, hasAdditionalMessages, isLoadingMessages, onLoadingMessages]);
+
+  const loadFollowingMessages = useCallback(async () => {
+    const lastMessage = conversation.getNewestMessage();
+
+    if (lastMessage) {
+      if (!isLastReceivedMessage(lastMessage, conversation)) {
+        onLoadingMessages(true);
+        // if the last loaded message is not the last of the conversation, we load the subsequent messages
+        await conversationRepository.getSubsequentMessages(conversation, lastMessage).then(() => {
+          onLoadingMessages(false);
+        });
+
+        nbOfMessages.current = itemsLength;
+      }
+    }
+  }, [conversation, conversationRepository, onLoadingMessages]);
+
+  useLayoutEffect(() => {
+    if (loadingMessages || !nbOfMessages.current) {
+      return;
+    }
+
+    const scrollElement = virtualizer.scrollElement;
+    if (!scrollElement) {
+      return;
+    }
+
+    virtualizer.scrollToIndex(itemsLength - nbOfMessages.current, {align: 'start'});
+  }, [itemsLength, loadingMessages]);
```

Comment:
can you add a comment explaining what is this useLayoutEffect trying to do here?